File: C#/LetterBudget.cs
using System.Collections.Generic;

namespace LetterPuzzle
{
  class LetterBudget
  {
    public LetterBudget(string word, LetterBudget letterBudget)
    {
      CharCounts.Capacity = letterBudget.Chars.Length;

      for (int i = 0; i < letterBudget.Chars.Length; i++)
      {
        int filterCharCount = 0;

        foreach (char wordChar in word)
        {
          if (wordChar == letterBudget.Chars[i])
          {
            filterCharCount++;
          }
        }

        if (letterBudget.CharCounts[i] != filterCharCount)
        {
          Chars += letterBudget.Chars[i];
          CharCounts.Add(letterBudget.CharCounts[i] - filterCharCount);
        }

        Count += letterBudget.CharCounts[i] - filterCharCount;
      }
    }

    public LetterBudget(string[] letters)
    {
      foreach (string letter in letters)
      {
        Chars += letter[0];
        Count += letter.Length;
        CharCounts.Add(letter.Length);
      }
    }


    public string Chars { get; private set; } = "";
    public List<int> CharCounts { get; private set; } = new List<int>();
    public int Count { get; private set; }
  }
}

File: Java/src/de/letterpuzzle/LetterBudget.java
package de.letterpuzzle;

import java.util.ArrayList;

public class LetterBudget {
    private String chars = "";
    private final ArrayList<Integer> charCounts;
    private int count = 0;

    public String getChars() {
        return chars;
    }

    public ArrayList<Integer> getCharCounts() {
        return charCounts;
    }

    public int getCount() {
        return count;
    }

    public LetterBudget(String word, LetterBudget letterBudget)
    {
        charCounts = new ArrayList<>(letterBudget.chars.length());

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < letterBudget.chars.length(); i++)
        {
            int filterCharCount = 0;

            for (char wordChar : word.toCharArray())
            {
                if (wordChar == letterBudget.getChars().charAt(i))
                {
                    filterCharCount++;
                }
            }

            if (letterBudget.getCharCounts().get(i) != filterCharCount)
            {
                sb.append(letterBudget.getChars().charAt(i));
                charCounts.add(letterBudget.getCharCounts().get(i) - filterCharCount);
            }

            count += letterBudget.getCharCounts().get(i) - filterCharCount;
        }

        chars =  sb.toString();
    }

    public LetterBudget(String[] letters)
    {
        charCounts = new ArrayList<>();

        StringBuilder sb = new StringBuilder();
        for (String letter : letters)
        {
            sb.append(letter.charAt(0));
            count += letter.length();
            charCounts.add(letter.length());
        }
        chars =  sb.toString();
    }
}
